src/rust-crypto/dehydration.ts

```diff
@@ -0,0 +1,262 @@
+        for (;;) {
+            const path = encodeUri("/dehydrated_device/$device_id/events", {
+                $device_id: dehydratedDeviceResp.device_id,
+            });
+            const eventResp: IDehydratedDeviceEventsResp = await this.http.authedRequest<IDehydratedDeviceEventsResp>(
+                Method.Post,
+                path,
+                undefined,
+                nextBatch ? { next_batch: nextBatch } : {},
+                {
+                    prefix: UnstablePrefix,
+                },
+            );
+
+            if (eventResp.events.length == 0) {
+                break;
+            }
+            toDeviceCount += eventResp.events.length;
+            nextBatch = eventResp.next_batch;
+            const roomKeyInfos: RustSdkCryptoJs.RoomKeyInfo[] = await rehydratedDevice.receiveEvents(
+                JSON.stringify(eventResp.events),
+            );
+            roomKeyCount += roomKeyInfos.length;
+
+            // FIXME: is this actually needed?  It looks like the OlmMachine
+            // automatically re-tries decryption
+            await this.rustCrypto.onRoomKeysUpdated(roomKeyInfos);
+        }
+        this.logger.info(`dehydration: received ${roomKeyCount} room keys from ${toDeviceCount} to-device events`);
+
+        return true;
+    }
+
+    /**
+     * Creates and uploads a new dehydrated device
+     *
+     * Creates and stores a new key in secret storage if none is available.
+     */
+    public async createAndUploadDehydratedDevice(): Promise<void> {
+        await this.getKey(true);
+
+        // FIXME: should raise error if server doesn't support dehydration?
+
+        const dehydratedDevice = await this.olmMachine.dehydratedDevices().create();
+        // FIXME: should the device display name be localised? passed as a
+        // parameter?
+        const request = await dehydratedDevice.keysForUpload("Dehydrated device", this.key!);
+
+        await this.outgoingRequestProcessor.makeOutgoingRequest(request);
+
+        this.logger.info("dehydration: uploaded device");
+    }
+
+    /**
+     * Schedule periodic creation of dehydrated devices
+     *
+     * @param interval - the time to wait between creating dehydrated devices
+     * @param delay - how long to wait before creating the first dehydrated device.
+     *     Defaults to creating the device immediately.
+     */
+    public async scheduleDeviceDehydration(interval: number, delay?: number): Promise<void> {
+        // cancel any previously-scheduled tasks
+        this.stop();
+
+        if (delay) {
+            this.timeoutId = setTimeout(() => {
+                this.scheduleDeviceDehydration(interval);
+                this.timeoutId = undefined;
+            }, delay);
+        } else {
+            await this.createAndUploadDehydratedDevice();
+            // FIXME: should we randomize the time a bit?
+            this.intervalId = setInterval(this.createAndUploadDehydratedDevice.bind(this), interval);
```

Comment:
we normally use an arrow function rather than an explicit `bind`:

```suggestion
            this.intervalId = setInterval(() => this.createAndUploadDehydratedDevice(), interval);
```

Reply:
Also, again, please add an explicit `.catch`